src/core/ext/transport/chaotic_good/server/chaotic_good_server.cc

```diff
@@ -0,0 +1,370 @@
+    : server_(server),
+      args_(args),
+      config_(args_),
+      event_engine_(grpc_event_engine::experimental::CreateEventEngine()) {}
+
+ChaoticGoodServerListener::~ChaoticGoodServerListener() {}
+
+absl::StatusOr<int> ChaoticGoodServerListener::Bind(const char* addr) {
+  EventEngine::Listener::AcceptCallback accept_cb =
+      [self = shared_from_this()](std::unique_ptr<EventEngine::Endpoint> ep,
+                                  MemoryAllocator memory_allocator) {
+        MutexLock lock(&self->mu_);
+        self->connection_ = std::make_shared<ActiveConnection>(self);
+        self->connection_->Start(std::move(ep));
+      };
+  auto shutdown_cb = [](absl::Status status) {
+    if (!status.ok()) {
+      gpr_log(GPR_ERROR, "Error server accept connection failed: %s",
+              StatusToString(status).c_str());
+    }
+  };
+  GPR_ASSERT(event_engine_ != nullptr);
+  auto ee_listener = event_engine_->CreateListener(
+      std::move(accept_cb), std::move(shutdown_cb), config_,
+      std::make_unique<grpc_core::MemoryQuota>("chaotic_good_server_listener"));
+  if (!ee_listener.ok()) {
+    return ee_listener.status();
+  }
+  ee_listener_ = std::move(ee_listener.value());
+  auto resolved_addr =
+      grpc_event_engine::experimental::URIToResolvedAddress(addr);
+  GPR_ASSERT(resolved_addr.ok());
+  if (!resolved_addr.ok()) {
+    return resolved_addr.status();
+  }
+  auto port_num = ee_listener_->Bind(resolved_addr.value());
+  if (!port_num.ok()) {
+    return port_num.status();
+  }
+  server_->AddListener(OrphanablePtr<Server::ListenerInterface>(this));
+  return port_num;
+}
+
+absl::Status ChaoticGoodServerListener::StartListening() {
+  GPR_ASSERT(ee_listener_ != nullptr);
+  auto status = ee_listener_->Start();
+  return status;
+}
+
+ChaoticGoodServerListener::ActiveConnection::ActiveConnection(
+    std::shared_ptr<ChaoticGoodServerListener> listener)
+    : listener_(listener),
+      memory_allocator_(
+          ResourceQuota::Default()->memory_quota()->CreateMemoryAllocator(
+              "connection")),
+      arena_(MakeScopedArena(1024, &memory_allocator_)),
+      context_(arena_.get()),
+      hpack_compressor_(std::make_unique<HPackCompressor>()),
+      hpack_parser_(std::make_unique<HPackParser>()),
+      connection_deadline_(Duration::Seconds(5)) {}
+
+ChaoticGoodServerListener::ActiveConnection::~ActiveConnection() {
+  listener_.reset();
+  if (receive_settings_activity_ != nullptr) receive_settings_activity_.reset();
+}
+
+void ChaoticGoodServerListener::ActiveConnection::Start(
+    std::unique_ptr<EventEngine::Endpoint> endpoint) {
+  GPR_ASSERT(handshaking_state_ == nullptr);
+  handshaking_state_ = std::make_shared<HandshakingState>(shared_from_this());
+  handshaking_state_->Start(std::move(endpoint));
+}
+
+void ChaoticGoodServerListener::ActiveConnection::GenerateConnectionID() {
+  std::string random_string;
+  int random_length = 8;
+  std::random_device rd;
```

Comment:
Use an `absl::BitGen` member variable guarded by a mutex instead - it'll be much lighter-weight than going to the OS for entropy and initializing the mt19937 state machine every time.